### PR TITLE
Simplify base rom verification and byteswapping

### PR DESCRIPTION
**Pull request by mzxrules**
Rom patching has been changed to where the program now patches a file called base.z64. this file is created whenever selecting a rom with the open file dialog. 

When starting up the program, base.z64 is checked to see if it exists in the roms folder and is a valid MM US rom with the correct byte ordering. If this passes, it will become the input rom for the program.

When selecting a base rom through the open button, the selected rom will be copied to base.z64 as long as it passes the validation check, or it will be byteswapped and copied over.

### CODE REVIEW

File: fMain.cs [hunk 1]
```diff
@@ -644,6 +657,72 @@ private void UpdateSettingsFromString(string settings)
...
+                    return false;
+                }
+            }
+            else if (validateResult == ValidateRomResult.Swap32 || validateResult == ValidateRomResult.Swap16)
+            {
+                byte[] data = new byte[0x2000000];
+                using (BinaryReader br = new BinaryReader(File.OpenRead(file)))
+                {
+                    br.Read(data, 0, data.Length);
+                }
```
Comment (kskjer) [inline fMain.cs]: Would it be simpler to use `var data = File.ReadAllBytes(file)` here?

File: fMain.cs [hunk 2]
```diff
@@ -644,6 +657,72 @@ private void UpdateSettingsFromString(string settings)
...
+                    {
+                        byte tmp = data[i];
+                        data[i] = data[i + 1];
+                        data[i + 1] = tmp;
+                    }
+                }
+                using (BinaryWriter br = new BinaryWriter(File.Create(BaseRomDirectory)))
+                {
+                    br.Write(data);
+                }
```
Comment (kskjer) [inline fMain.cs]: Simpler to use `File.WriteAllBytes` here?

**Comment by kskjer** [comment]
Personally I would decouple file I/O and byte swapping. Also, the swapping code can be simplified.

```csharp

        static byte[] Swap(byte[] input, byte xor)
        {
            if (xor == 0)
                return input;

            var output = new byte[input.Length];

            for (var i = 0; i < input.Length; i++)
                output[i] = input[i ^ xor];

            return output;
        }
```

So then the whole `TryCopyToBaseRomLocation` method can pretty much be replaced with:

```csharp
private bool TryCopyToBaseRomLocation(string file, ValidateRomResult validateResult)
{
    File.WriteAllBytes(
        file,
        Swap(
            File.ReadAllBytes(file),
            validateResult == ValidateRomResult.Swap32 ? 3 :
            validateResult == ValidateRomResult.Swap16 ? 1 : 0
        )
    );
    ...
}
```

File: fMain.cs [hunk 3]
```diff
@@ -644,6 +657,56 @@ private void UpdateSettingsFromString(string settings)
...
+        private void InitializeSelectedRom()
+        {
+            if(ValidateRom(BaseRomDirectory) == ValidateRomResult.ValidFile)
+            {
+                Settings.InputROMFilename = BaseRomDirectory;
+                tROMName.Text = Settings.InputROMFilename;
+            }
+        }
+
+        private bool TryCopyToBaseRomLocation(string file, ValidateRomResult validateResult)
```
[the remark this answers is not in the document]
Comment (mzxrules) [reply fMain.cs]: I can understand putting SwapByteOrder somewhere else, but not sure where the appropriate place would be. I don't exactly understand the logic behind moving TryCopyToBaseRomLocation though

[the remark this answers is not in the document]
Comment (Kjelli) [reply fMain.cs]: `fMain.cs` represents a form - the form itself has functionality that is triggered from the ui, but is not part of the ui. The logic would be, separation of concerns. You wouldn't want to build everything in a single class, would you?

File: ValidateRomResult.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿namespace MMRando
```
Comment (Kjelli) [inline ValidateRomResult.cs]: I know, subfolders are a huge pain to work with - let's just put random ass stuff in the root folder. Structure is overrated anyway.

That being said, could you move this to models folder? 😏

Comment (mzxrules) [reply ValidateRomResult.cs]: I'd rather leave it in root over putting it into a completely unrelated folder where it can get lost

Comment (Kjelli) [reply ValidateRomResult.cs]: Don't you just hate it when your files just get completely lost...

... in a subfolder

**Comment by mzxrules** [comment]
#55

**Pull request closed without merging**